tla/consensus/ccfraft.tla

```diff
@@ -456,28 +456,49 @@ PlausibleSucessorNodes(i) ==
         highestMatchServers == {n \in activeServers : \A m \in activeServers : matchIndex[i][n] >= matchIndex[i][m]}
     IN {n \in highestMatchServers : \A m \in highestMatchServers: HighestConfigurationWithNode(i, n) >= HighestConfigurationWithNode(i, m)}
 
+\* Generate initial state for a given startNode
+StartState(startNode, servers) ==
+    /\ reconfigurationCount = 0
+    /\ removedFromConfiguration = {}
+    /\ configurations = [ i \in servers |-> IF i = startNode THEN (1 :> {startNode}) ELSE << >>]
+    /\ currentTerm = [i \in servers |-> IF i = startNode THEN 2 ELSE 0]
+    /\ state       = [i \in servers |-> IF i = startNode THEN Leader ELSE None]
+    /\ votedFor    = [i \in servers |-> Nil]
+    /\ log         = [i \in servers |-> IF i = startNode
+                                        THEN << [term |-> 2, contentType |-> TypeReconfiguration, configuration |-> {startNode}],
+                                                [term |-> 2, contentType |-> TypeSignature] >>
+                                        ELSE << >>]
+    /\ commitIndex  = [i \in servers |-> IF i = startNode THEN 2 ELSE 0]
+    /\ committableIndices  = [i \in servers |-> {}]
+
+\* Generate initial state for a startNodes configuration
+JoinedState(startNodes, servers) ==
+    /\ reconfigurationCount = 0
+    /\ removedFromConfiguration = {}
+    /\ \E startNode \in startNodes:
+        /\ configurations = [ i \in servers |-> IF i \in startNodes  THEN (3 :> startNodes) ELSE << >>]
+        /\ currentTerm = [i \in servers |-> IF i \in startNodes THEN 2 ELSE 0]
+        /\ state       = [i \in servers |-> IF i = startNode THEN Leader ELSE IF i \in startNodes THEN Follower ELSE None]
+        /\ votedFor    = [i \in servers |-> Nil]
+        /\ log         = [i \in servers |-> IF i \in startNodes
+                                            THEN << [term |-> 2, contentType |-> TypeReconfiguration, configuration |-> {startNode}],
+                                                    [term |-> 2, contentType |-> TypeSignature],
+                                                    [term |-> 2, contentType |-> TypeReconfiguration, configuration |-> startNodes],
+                                                    [term |-> 2, contentType |-> TypeSignature] >>
+                                            ELSE << >>]
+        /\ commitIndex  = [i \in servers |-> IF i \in startNodes THEN 4 ELSE 0]
+        /\ committableIndices  = [i \in servers |-> {}]
+
```

Comment:
Given the similarities, `StartNode` and `JoinedState` could be refactored into a single definition parameterized by the log.

```tla
InitLogConfigServerVars(startNodes, log) ==
    /\ reconfigurationCount = 0
    /\ removedFromConfiguration = {}
    /\ \E startNode \in startNodes:
        /\ configurations = [ i \in Servers |-> IF i \in startNodes  THEN (Len(log)-1 :> startNodes) ELSE << >>]
        /\ currentTerm = [i \in Servers |-> IF i \in startNodes THEN 2 ELSE 0]
        /\ state       = [i \in Servers |-> IF i = startNode THEN Leader ELSE IF i \in startNodes THEN Follower ELSE None]
        /\ votedFor    = [i \in Servers |-> Nil]
        /\ log         = [i \in Servers |-> IF i \in startNodes
                                            THEN log
                                            ELSE << >>]
        /\ commitIndex  = [i \in Servers |-> IF i \in startNodes THEN Len(log) ELSE 0]
        /\ committableIndices  = [i \in Servers |-> {}]

Reply:
Yes, but there's only two, and that also makes it unsafe because it's then easy to make a call to InitLogConfigServerVars that generates an unrealistic startNodes, log pair. If it could be made private, and then JoinedState and StartNode were the only two public ways to call into it, it'd be good, but I haven't found a way to do that.

Reply:
So I had a go, but it turns out not to work, because to write the log in the first place, you need to take out the startNode, and you end up having to write a bunch of extra checks to make sure the log is half plausible.

So these are the two correct start states for now, and if we really need to make new ones, we will revisit this. But the next order of business (#5822) is to get the trace validation to work for all scenarios. Extending for more initial states will wait for now.

```
_InitialStateWithLog(initialLog, servers) ==
    /\ Len(initialLog) >= 2
    \* Log must end with a config and a signature
    /\ initialLog[Len(initialLog) - 1].contentType = TypeReconfiguration
    /\ initialLog[Len(initialLog)].contentType = TypeSignature
    \* Log must end with a single node config and a signature
    /\ initialLog[1].contentType = TypeReconfiguration
    /\ initialLog[2].contentType = TypeSignature
    /\ DOMAIN initialLog[1].configuration = {1}
    /\ Cardinality(initialLog[1].configuration[1]) = 1
    /\ reconfigurationCount = 0
    /\ removedFromConfiguration = {}
    /\ LET startNodes == initialLog[Len(initialLog) - 1].configurations[Max(DOMAIN configurations)]
       IN \E startNode \in startNodes:
        /\ configurations = [ i \in servers |-> IF i \in startNodes  THEN (3 :> startNodes) ELSE << >>]
        /\ currentTerm = [i \in servers |-> IF i \in startNodes THEN StartTerm ELSE 0]
        /\ state       = [i \in servers |-> IF i = startNode THEN Leader ELSE IF i \in startNodes THEN Follower ELSE None]
        /\ votedFor    = [i \in servers |-> Nil]
        /\ log         = [i \in servers |-> IF i \in startNodes THEN initialLog ELSE << >>]
        /\ commitIndex  = [i \in servers |-> IF i \in startNodes THEN Len(log[i]) ELSE 0]
        /\ committableIndices  = [i \in servers |-> {}]
```